File: app/src/test/java/eu/kanade/tachiyomi/BackupTest.java
package eu.kanade.tachiyomi;

import android.app.Application;
import android.os.Build;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import java.util.ArrayList;
import java.util.List;

import eu.kanade.tachiyomi.data.backup.BackupManager;
import eu.kanade.tachiyomi.data.database.DatabaseHelper;
import eu.kanade.tachiyomi.data.database.models.Category;
import eu.kanade.tachiyomi.data.database.models.Chapter;
import eu.kanade.tachiyomi.data.database.models.Manga;
import eu.kanade.tachiyomi.data.database.models.MangaCategory;
import eu.kanade.tachiyomi.data.database.models.MangaSync;

import static org.assertj.core.api.Assertions.assertThat;

@Config(constants = BuildConfig.class, sdk = Build.VERSION_CODES.LOLLIPOP)
@RunWith(CustomRobolectricGradleTestRunner.class)
public class BackupTest {

    DatabaseHelper db;
    BackupManager backupManager;
    Gson gson;
    JsonObject root;

    @Before
    public void setup() {
        Application app = RuntimeEnvironment.application;
        db = new DatabaseHelper(app);
        backupManager = new BackupManager(db);
        gson = new Gson();
        root = new JsonObject();
    }

    @Test
    public void testRestoreCategory() {
        String catName = "cat";
        root = createRootJson(null, toJson(createCategories(catName)));
        backupManager.restoreFromJson(root);

        List<Category> dbCats = db.getCategories().executeAsBlocking();
        assertThat(dbCats).hasSize(1);
        assertThat(dbCats.get(0).name).isEqualTo(catName);
    }

    @Test
    public void testRestoreEmptyCategory() {
        root = createRootJson(null, toJson(new ArrayList<>()));
        backupManager.restoreFromJson(root);
        List<Category> dbCats = db.getCategories().executeAsBlocking();
        assertThat(dbCats).isEmpty();
    }

    @Test
    public void testRestoreExistingCategory() {
        String catName = "cat";
        db.insertCategory(createCategory(catName)).executeAsBlocking();

        root = createRootJson(null, toJson(createCategories(catName)));
        backupManager.restoreFromJson(root);

        List<Category> dbCats = db.getCategories().executeAsBlocking();
        assertThat(dbCats).hasSize(1);
        assertThat(dbCats.get(0).name).isEqualTo(catName);
    }

    @Test
    public void testRestoreCategories() {
        root = createRootJson(null, toJson(createCategories("cat", "cat2", "cat3")));
        backupManager.restoreFromJson(root);

        List<Category> dbCats = db.getCategories().executeAsBlocking();
        assertThat(dbCats).hasSize(3);
    }

    @Test
    public void testRestoreExistingCategories() {
        db.insertCategories(createCategories("cat", "cat2")).executeAsBlocking();

        root = createRootJson(null, toJson(createCategories("cat", "cat2", "cat3")));
        backupManager.restoreFromJson(root);

        List<Category> dbCats = db.getCategories().executeAsBlocking();
        assertThat(dbCats).hasSize(3);
    }

    @Test
    public void testRestoreExistingCategoriesAlt() {
        db.insertCategories(createCategories("cat", "cat2", "cat3")).executeAsBlocking();

        root = createRootJson(null, toJson(createCategories("cat", "cat2")));
        backupManager.restoreFromJson(root);

        List<Category> dbCats = db.getCategories().executeAsBlocking();
        assertThat(dbCats).hasSize(3);
    }

    @Test
    public void testRestoreManga() {
        String mangaName = "title";
        List<Manga> mangas = createMangas(mangaName);
        List<JsonElement> elements = new ArrayList<>();
        for (Manga manga : mangas) {
            JsonObject entry = new JsonObject();
            entry.add("manga", toJson(manga));
            elements.add(entry);
        }
        root = createRootJson(toJson(elements), null);
        backupManager.restoreFromJson(root);

        List<Manga> dbMangas = db.getMangas().executeAsBlocking();
        assertThat(dbMangas).hasSize(1);
        assertThat(dbMangas.get(0).title).isEqualTo(mangaName);
    }

    @Test
    public void testRestoreExistingManga() {
        String mangaName = "title";
        Manga manga = createManga(mangaName);

        db.insertManga(manga).executeAsBlocking();

        List<JsonElement> elements = new ArrayList<>();
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        elements.add(entry);

        root = createRootJson(toJson(elements), null);
        backupManager.restoreFromJson(root);

        List<Manga> dbMangas = db.getMangas().executeAsBlocking();
        assertThat(dbMangas).hasSize(1);
    }

    @Test
    public void testRestoreExistingMangaWithUpdatedFields() {
        // Store a manga in db
        String mangaName = "title";
        String updatedThumbnailUrl = "updated thumbnail url";
        Manga manga = createManga(mangaName);
        manga.chapter_flags = 1024;
        manga.thumbnail_url = updatedThumbnailUrl;
        db.insertManga(manga).executeAsBlocking();

        // Add an entry for a new manga with different attributes
        manga = createManga(mangaName);
        manga.chapter_flags = 512;
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));

        // Append the entry to the backup list
        List<JsonElement> elements = new ArrayList<>();
        elements.add(entry);

        // Restore from json
        root = createRootJson(toJson(elements), null);
        backupManager.restoreFromJson(root);

        List<Manga> dbMangas = db.getMangas().executeAsBlocking();
        assertThat(dbMangas).hasSize(1);
        assertThat(dbMangas.get(0).thumbnail_url).isEqualTo(updatedThumbnailUrl);
        assertThat(dbMangas.get(0).chapter_flags).isEqualTo(512);
    }

    @Test
    public void testRestoreChaptersForManga() {
        // Create a manga and 3 chapters
        Manga manga = createManga("title");
        manga.id = 1L;
        List<Chapter> chapters = createChapters(manga, "1", "2", "3");

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("chapters", toJson(chapters));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), null);
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(1).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        List<Chapter> dbChapters = db.getChapters(dbManga).executeAsBlocking();
        assertThat(dbChapters).hasSize(3);
    }

    @Test
    public void testRestoreChaptersForExistingManga() {
        long mangaId = 3;
        // Create a manga and 3 chapters
        Manga manga = createManga("title");
        manga.id = mangaId;
        List<Chapter> chapters = createChapters(manga, "1", "2", "3");
        db.insertManga(manga).executeAsBlocking();

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("chapters", toJson(chapters));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), null);
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(mangaId).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        List<Chapter> dbChapters = db.getChapters(dbManga).executeAsBlocking();
        assertThat(dbChapters).hasSize(3);
    }

    @Test
    public void testRestoreExistingChaptersForExistingManga() {
        long mangaId = 5;
        // Store a manga and 3 chapters
        Manga manga = createManga("title");
        manga.id = mangaId;
        List<Chapter> chapters = createChapters(manga, "1", "2", "3");
        db.insertManga(manga).executeAsBlocking();
        db.insertChapters(chapters).executeAsBlocking();

        // The backup contains a existing chapter and a new one, so it should have 4 chapters
        chapters = createChapters(manga, "3", "4");

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("chapters", toJson(chapters));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), null);
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(mangaId).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        List<Chapter> dbChapters = db.getChapters(dbManga).executeAsBlocking();
        assertThat(dbChapters).hasSize(4);
    }

    @Test
    public void testRestoreCategoriesForManga() {
        // Create a manga
        Manga manga = createManga("title");

        // Create categories
        List<Category> categories = createCategories("cat1", "cat2", "cat3");

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("categories", toJson(createStringCategories("cat1")));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), toJson(categories));
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(1).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        assertThat(db.getCategoriesForManga(dbManga).executeAsBlocking())
                .hasSize(1)
                .contains(Category.create("cat1"))
                .doesNotContain(Category.create("cat2"));
    }

    @Test
    public void testRestoreCategoriesForExistingManga() {
        // Store a manga
        Manga manga = createManga("title");
        db.insertManga(manga).executeAsBlocking();

        // Create categories
        List<Category> categories = createCategories("cat1", "cat2", "cat3");

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("categories", toJson(createStringCategories("cat1")));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), toJson(categories));
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(1).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        assertThat(db.getCategoriesForManga(dbManga).executeAsBlocking())
                .hasSize(1)
                .contains(Category.create("cat1"))
                .doesNotContain(Category.create("cat2"));
    }

    @Test
    public void testRestoreMultipleCategoriesForManga() {
        // Create a manga
        Manga manga = createManga("title");

        // Create categories
        List<Category> categories = createCategories("cat1", "cat2", "cat3");

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("categories", toJson(createStringCategories("cat1", "cat3")));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), toJson(categories));
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(1).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        assertThat(db.getCategoriesForManga(dbManga).executeAsBlocking())
                .hasSize(2)
                .contains(Category.create("cat1"), Category.create("cat3"))
                .doesNotContain(Category.create("cat2"));
    }

    @Test
    public void testRestoreMultipleCategoriesForExistingMangaAndCategory() {
        // Store a manga and a category
        Manga manga = createManga("title");
        manga.id = 1L;
        db.insertManga(manga).executeAsBlocking();

        Category cat = createCategory("cat1");
        cat.id = 1;
        db.insertCategory(cat).executeAsBlocking();
        db.insertMangaCategory(MangaCategory.create(manga, cat)).executeAsBlocking();

        // Create categories
        List<Category> categories = createCategories("cat1", "cat2", "cat3");

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("categories", toJson(createStringCategories("cat1", "cat2")));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), toJson(categories));
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(1).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        assertThat(db.getCategoriesForManga(dbManga).executeAsBlocking())
                .hasSize(2)
                .contains(Category.create("cat1"), Category.create("cat2"))
                .doesNotContain(Category.create("cat3"));
    }

    @Test
    public void testRestoreSyncForManga() {
        // Create a manga and mangaSync
        Manga manga = createManga("title");
        manga.id = 1L;

        List<MangaSync> mangaSync = createMangaSync(manga, 1, 2, 3);

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("sync", toJson(mangaSync));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), null);
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(1).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        List<MangaSync> dbSync = db.getMangasSync(dbManga).executeAsBlocking();
        assertThat(dbSync).hasSize(3);
    }

    @Test
    public void testRestoreSyncForExistingManga() {
        long mangaId = 3;
        // Create a manga and 3 sync
        Manga manga = createManga("title");
        manga.id = mangaId;
        List<MangaSync> mangaSync = createMangaSync(manga, 1, 2, 3);
        db.insertManga(manga).executeAsBlocking();

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("sync", toJson(mangaSync));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), null);
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(mangaId).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        List<MangaSync> dbSync = db.getMangasSync(dbManga).executeAsBlocking();
        assertThat(dbSync).hasSize(3);
    }

    @Test
    public void testRestoreExistingSyncForExistingManga() {
        long mangaId = 5;
        // Store a manga and 3 sync
        Manga manga = createManga("title");
        manga.id = mangaId;
        List<MangaSync> mangaSync = createMangaSync(manga, 1, 2, 3);
        db.insertManga(manga).executeAsBlocking();
        db.insertMangasSync(mangaSync).executeAsBlocking();

        // The backup contains a existing sync and a new one, so it should have 4 sync
        mangaSync = createMangaSync(manga, 3, 4);

        // Add an entry for the manga
        JsonObject entry = new JsonObject();
        entry.add("manga", toJson(manga));
        entry.add("sync", toJson(mangaSync));

        // Append the entry to the backup list
        List<JsonElement> mangas = new ArrayList<>();
        mangas.add(entry);

        // Restore from json
        root = createRootJson(toJson(mangas), null);
        backupManager.restoreFromJson(root);

        Manga dbManga = db.getManga(mangaId).executeAsBlocking();
        assertThat(dbManga).isNotNull();

        List<MangaSync> dbSync = db.getMangasSync(dbManga).executeAsBlocking();
        assertThat(dbSync).hasSize(4);
    }

    private JsonObject createRootJson(JsonElement mangas, JsonElement categories) {
        JsonObject root = new JsonObject();
        if (mangas != null)
            root.add("mangas", mangas);
        if (categories != null)
            root.add("categories", categories);
        return root;
    }

    private Category createCategory(String name) {
        Category c = new Category();
        c.name = name;
        return c;
    }

    private List<Category> createCategories(String... names) {
        List<Category> cats = new ArrayList<>();
        for (String name : names) {
            cats.add(createCategory(name));
        }
        return cats;
    }

    private List<String> createStringCategories(String... names) {
        List<String> cats = new ArrayList<>();
        for (String name : names) {
            cats.add(name);
        }
        return cats;
    }

    private Manga createManga(String title) {
        Manga m = new Manga();
        m.title = title;
        m.author = "";
        m.artist = "";
        m.thumbnail_url = "";
        m.genre = "a list of genres";
        m.description = "long description";
        m.url = "url to manga";
        m.favorite = true;
        m.source = 1;
        return m;
    }

    private List<Manga> createMangas(String... titles) {
        List<Manga> mangas = new ArrayList<>();
        for (String title : titles) {
            mangas.add(createManga(title));
        }
        return mangas;
    }

    private Chapter createChapter(Manga manga, String url) {
        Chapter c = Chapter.create();
        c.url = url;
        c.name = url;
        c.manga_id = manga.id;
        return c;
    }

    private List<Chapter> createChapters(Manga manga, String... urls) {
        List<Chapter> chapters = new ArrayList<>();
        for (String url : urls) {
            chapters.add(createChapter(manga, url));
        }
        return chapters;
    }

    private MangaSync createMangaSync(Manga manga, int syncId) {
        MangaSync m = MangaSync.create();
        m.manga_id = manga.id;
        m.sync_id = syncId;
        m.title = "title";
        return m;
    }

    private List<MangaSync> createMangaSync(Manga manga, Integer... syncIds) {
        List<MangaSync> ms = new ArrayList<>();
        for (int title : syncIds) {
            ms.add(createMangaSync(manga, title));
        }
        return ms;
    }

    private JsonElement toJson(Object element) {
        return gson.toJsonTree(element);
    }

}

File: app/src/test/java/eu/kanade/tachiyomi/data/backup/BackupTest.kt
package eu.kanade.tachiyomi.data.backup

import android.app.Application
import android.content.Context
import android.os.Build
import com.github.salomonbrys.kotson.fromJson
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import eu.kanade.tachiyomi.BuildConfig
import eu.kanade.tachiyomi.CustomRobolectricGradleTestRunner
import eu.kanade.tachiyomi.data.backup.legacy.LegacyBackupManager
import eu.kanade.tachiyomi.data.backup.legacy.models.Backup
import eu.kanade.tachiyomi.data.backup.legacy.models.DHistory
import eu.kanade.tachiyomi.data.database.DatabaseHelper
import eu.kanade.tachiyomi.data.database.models.Category
import eu.kanade.tachiyomi.data.database.models.Chapter
import eu.kanade.tachiyomi.data.database.models.ChapterImpl
import eu.kanade.tachiyomi.data.database.models.Manga
import eu.kanade.tachiyomi.data.database.models.MangaImpl
import eu.kanade.tachiyomi.data.database.models.TrackImpl
import eu.kanade.tachiyomi.source.SourceManager
import eu.kanade.tachiyomi.source.online.HttpSource
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.RETURNS_DEEP_STUBS
import org.mockito.Mockito.`when`
import org.mockito.Mockito.anyLong
import org.mockito.Mockito.mock
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config
import rx.Observable
import uy.kohesive.injekt.Injekt
import uy.kohesive.injekt.api.InjektModule
import uy.kohesive.injekt.api.InjektRegistrar
import uy.kohesive.injekt.api.addSingleton

/**
 * Test class for the [LegacyBackupManager].
 * Note that this does not include the backup create/restore services.
 */
@Config(constants = BuildConfig::class, sdk = [Build.VERSION_CODES.LOLLIPOP])
@RunWith(CustomRobolectricGradleTestRunner::class)
class BackupTest {
    // Create root object
    var root = JsonObject()

    // Create information object
    var information = JsonObject()

    // Create manga array
    var mangaEntries = JsonArray()

    // Create category array
    var categoryEntries = JsonArray()

    lateinit var app: Application
    lateinit var context: Context
    lateinit var source: HttpSource

    lateinit var legacyBackupManager: LegacyBackupManager

    lateinit var db: DatabaseHelper

    @Before
    fun setup() {
        app = RuntimeEnvironment.application
        context = app.applicationContext
        legacyBackupManager = LegacyBackupManager(context, 2)
        db = legacyBackupManager.databaseHelper

        // Mock the source manager
        val module = object : InjektModule {
            override fun InjektRegistrar.registerInjectables() {
                addSingleton(mock(SourceManager::class.java, RETURNS_DEEP_STUBS))
            }
        }
        Injekt.importModule(module)

        source = mock(HttpSource::class.java)
        `when`(legacyBackupManager.sourceManager.get(anyLong())).thenReturn(source)

        root.add(Backup.MANGAS, mangaEntries)
        root.add(Backup.CATEGORIES, categoryEntries)

        clearJson()
    }

    /**
     * Test that checks if no crashes when no categories in library.
     */
    @Test
    fun testRestoreEmptyCategory() {
        // Create backup of empty database
        legacyBackupManager.backupCategories(categoryEntries)

        // Restore Json
        legacyBackupManager.restoreCategories(categoryEntries)

        // Check if empty
        val dbCats = db.getCategories().executeAsBlocking()
        assertThat(dbCats).isEmpty()
    }

    /**
     * Test to check if single category gets restored
     */
    @Test
    fun testRestoreSingleCategory() {
        // Create category and add to json
        val category = addSingleCategory("category")

        // Restore Json
        legacyBackupManager.restoreCategories(categoryEntries)

        // Check if successful
        val dbCats = legacyBackupManager.databaseHelper.getCategories().executeAsBlocking()
        assertThat(dbCats).hasSize(1)
        assertThat(dbCats[0].name).isEqualTo(category.name)
    }

    /**
     * Test to check if multiple categories get restored.
     */
    @Test
    fun testRestoreMultipleCategories() {
        // Create category and add to json
        val category = addSingleCategory("category")
        val category2 = addSingleCategory("category2")
        val category3 = addSingleCategory("category3")
        val category4 = addSingleCategory("category4")
        val category5 = addSingleCategory("category5")

        // Insert category to test if no duplicates on restore.
        db.insertCategory(category).executeAsBlocking()

        // Restore Json
        legacyBackupManager.restoreCategories(categoryEntries)

        // Check if successful
        val dbCats = legacyBackupManager.databaseHelper.getCategories().executeAsBlocking()
        assertThat(dbCats).hasSize(5)
        assertThat(dbCats[0].name).isEqualTo(category.name)
        assertThat(dbCats[1].name).isEqualTo(category2.name)
        assertThat(dbCats[2].name).isEqualTo(category3.name)
        assertThat(dbCats[3].name).isEqualTo(category4.name)
        assertThat(dbCats[4].name).isEqualTo(category5.name)
    }

    /**
     * Test if restore of manga is successful
     */
    @Test
    fun testRestoreManga() {
        // Add manga to database
        val manga = getSingleManga("One Piece")
        manga.viewer = 3
        manga.id = db.insertManga(manga).executeAsBlocking().insertedId()

        var favoriteManga = legacyBackupManager.databaseHelper.getFavoriteMangas().executeAsBlocking()
        assertThat(favoriteManga).hasSize(1)
        assertThat(favoriteManga[0].viewer).isEqualTo(3)

        // Update json with all options enabled
        mangaEntries.add(legacyBackupManager.backupMangaObject(manga, 1))

        // Change manga in database to default values
        val dbManga = getSingleManga("One Piece")
        dbManga.id = manga.id
        db.insertManga(dbManga).executeAsBlocking()

        favoriteManga = legacyBackupManager.databaseHelper.getFavoriteMangas().executeAsBlocking()
        assertThat(favoriteManga).hasSize(1)
        assertThat(favoriteManga[0].viewer).isEqualTo(0)

        // Restore local manga
        legacyBackupManager.restoreMangaNoFetch(manga, dbManga)

        // Test if restore successful
        favoriteManga = legacyBackupManager.databaseHelper.getFavoriteMangas().executeAsBlocking()
        assertThat(favoriteManga).hasSize(1)
        assertThat(favoriteManga[0].viewer).isEqualTo(3)

        // Clear database to test manga fetch
        clearDatabase()

        // Test if successful
        favoriteManga = legacyBackupManager.databaseHelper.getFavoriteMangas().executeAsBlocking()
        assertThat(favoriteManga).hasSize(0)

        // Restore Json
        // Create JSON from manga to test parser
        val json = legacyBackupManager.parser.toJsonTree(manga)
        // Restore JSON from manga to test parser
        val jsonManga = legacyBackupManager.parser.fromJson<MangaImpl>(json)

        // Restore manga with fetch observable
        val networkManga = getSingleManga("One Piece")
        networkManga.description = "This is a description"
        `when`(source.fetchMangaDetails(jsonManga)).thenReturn(Observable.just(networkManga))

        runBlocking {
            legacyBackupManager.fetchManga(source, jsonManga)

            // Check if restore successful
            val dbCats = legacyBackupManager.databaseHelper.getFavoriteMangas().executeAsBlocking()
            assertThat(dbCats).hasSize(1)
            assertThat(dbCats[0].viewer).isEqualTo(3)
            assertThat(dbCats[0].description).isEqualTo("This is a description")
        }
    }

    /**
     * Test if chapter restore is successful
     */
    @Test
    fun testRestoreChapters() {
        // Insert manga
        val manga = getSingleManga("One Piece")
        manga.id = legacyBackupManager.databaseHelper.insertManga(manga).executeAsBlocking().insertedId()

        // Create restore list
        val chapters = mutableListOf<Chapter>()
        for (i in 1..8) {
            val chapter = getSingleChapter("Chapter $i")
            chapter.read = true
            chapters.add(chapter)
        }

        // Check parser
        val chaptersJson = legacyBackupManager.parser.toJsonTree(chapters)
        val restoredChapters = legacyBackupManager.parser.fromJson<List<ChapterImpl>>(chaptersJson)

        // Fetch chapters from upstream
        // Create list
        val chaptersRemote = mutableListOf<Chapter>()
        (1..10).mapTo(chaptersRemote) { getSingleChapter("Chapter $it") }
        `when`(source.fetchChapterList(manga)).thenReturn(Observable.just(chaptersRemote))

        runBlocking {
            legacyBackupManager.restoreChapters(source, manga, restoredChapters)

            val dbCats = legacyBackupManager.databaseHelper.getChapters(manga).executeAsBlocking()
            assertThat(dbCats).hasSize(10)
            assertThat(dbCats[0].read).isEqualTo(true)
        }
    }

    /**
     * Test to check if history restore works
     */
    @Test
    fun restoreHistoryForManga() {
        val manga = getSingleManga("One Piece")
        manga.id = legacyBackupManager.databaseHelper.insertManga(manga).executeAsBlocking().insertedId()

        // Create chapter
        val chapter = getSingleChapter("Chapter 1")
        chapter.manga_id = manga.id
        chapter.read = true
        chapter.id = legacyBackupManager.databaseHelper.insertChapter(chapter).executeAsBlocking().insertedId()

        val historyJson = getSingleHistory(chapter)

        val historyList = mutableListOf<DHistory>()
        historyList.add(historyJson)

        // Check parser
        val historyListJson = legacyBackupManager.parser.toJsonTree(historyList)
        val history = legacyBackupManager.parser.fromJson<List<DHistory>>(historyListJson)

        // Restore categories
        legacyBackupManager.restoreHistoryForManga(history)

        val historyDB = legacyBackupManager.databaseHelper.getHistoryByMangaId(manga.id!!).executeAsBlocking()
        assertThat(historyDB).hasSize(1)
        assertThat(historyDB[0].last_read).isEqualTo(1000)
    }

    /**
     * Test to check if tracking restore works
     */
    @Test
    fun restoreTrackForManga() {
        // Create mangas
        val manga = getSingleManga("One Piece")
        val manga2 = getSingleManga("Bleach")
        manga.id = legacyBackupManager.databaseHelper.insertManga(manga).executeAsBlocking().insertedId()
        manga2.id = legacyBackupManager.databaseHelper.insertManga(manga2).executeAsBlocking().insertedId()

        // Create track and add it to database
        // This tests duplicate errors.
        val track = getSingleTrack(manga)
        track.last_chapter_read = 5
        legacyBackupManager.databaseHelper.insertTrack(track).executeAsBlocking()
        var trackDB = legacyBackupManager.databaseHelper.getTracks(manga).executeAsBlocking()
        assertThat(trackDB).hasSize(1)
        assertThat(trackDB[0].last_chapter_read).isEqualTo(5)
        track.last_chapter_read = 7

        // Create track for different manga to test track not in database
        val track2 = getSingleTrack(manga2)
        track2.last_chapter_read = 10

        // Check parser and restore already in database
        var trackList = listOf(track)
        // Check parser
        var trackListJson = legacyBackupManager.parser.toJsonTree(trackList)
        var trackListRestore = legacyBackupManager.parser.fromJson<List<TrackImpl>>(trackListJson)
        legacyBackupManager.restoreTrackForManga(manga, trackListRestore)

        // Assert if restore works.
        trackDB = legacyBackupManager.databaseHelper.getTracks(manga).executeAsBlocking()
        assertThat(trackDB).hasSize(1)
        assertThat(trackDB[0].last_chapter_read).isEqualTo(7)

        // Check parser and restore already in database with lower chapter_read
        track.last_chapter_read = 5
        trackList = listOf(track)
        legacyBackupManager.restoreTrackForManga(manga, trackList)

        // Assert if restore works.
        trackDB = legacyBackupManager.databaseHelper.getTracks(manga).executeAsBlocking()
        assertThat(trackDB).hasSize(1)
        assertThat(trackDB[0].last_chapter_read).isEqualTo(7)

        // Check parser and restore, track not in database
        trackList = listOf(track2)

        // Check parser
        trackListJson = legacyBackupManager.parser.toJsonTree(trackList)
        trackListRestore = legacyBackupManager.parser.fromJson<List<TrackImpl>>(trackListJson)
        legacyBackupManager.restoreTrackForManga(manga2, trackListRestore)

        // Assert if restore works.
        trackDB = legacyBackupManager.databaseHelper.getTracks(manga2).executeAsBlocking()
        assertThat(trackDB).hasSize(1)
        assertThat(trackDB[0].last_chapter_read).isEqualTo(10)
    }

    private fun clearJson() {
        root = JsonObject()
        information = JsonObject()
        mangaEntries = JsonArray()
        categoryEntries = JsonArray()
    }

    private fun addSingleCategory(name: String): Category {
        val category = Category.create(name)
        val catJson = legacyBackupManager.parser.toJsonTree(category)
        categoryEntries.add(catJson)
        return category
    }

    private fun clearDatabase() {
        db.deleteMangas().executeAsBlocking()
        db.deleteHistory().executeAsBlocking()
    }

    private fun getSingleHistory(chapter: Chapter): DHistory {
        return DHistory(chapter.url, 1000)
    }

    private fun getSingleTrack(manga: Manga): TrackImpl {
        val track = TrackImpl()
        track.title = manga.title
        track.manga_id = manga.id!!
        track.sync_id = 1
        return track
    }

    private fun getSingleManga(title: String): MangaImpl {
        val manga = MangaImpl()
        manga.source = 1
        manga.title = title
        manga.url = "/manga/$title"
        manga.favorite = true
        return manga
    }

    private fun getSingleChapter(name: String): ChapterImpl {
        val chapter = ChapterImpl()
        chapter.name = name
        chapter.url = "/read-online/$name-page-1.html"
        return chapter
    }
}
